Add support for custom auth token expire date in UI and API

michalkleiner opened the pull request:
### Description:

When adding personal auth token, it's now possible to define an expire date, after which the token will be rejected by Matomo.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

michalkleiner committed:
Add support for custom auth token expire date in UI and API

sgiehl commented on plugins/UsersManager/Controller.php:
```diff
@@ -387,37 +387,67 @@ public function addNewToken()
 
-        $noDescription = false;
+        $invalidExpireDate = null;
+        if (!empty($_POST['token_expire_date'])) {
+            $tokenExpireDate = \Piwik\Request::fromRequest()->getStringParameter('token_expire_date');
```
What is that check supposed to do? Do you want to ensure it's sent by POST? Currently if the value would be sent by GET and POST, it would actually use the GET value. The already existing code might already haven been broken 🤔 
You could simply use `\Piwik\Request::fromPost()` for that. 
And it might be beneficial to only call `\Piwik\Request::fromPost()` once and assign it to a variable that can be reused.

michalkleiner replied:
I went by the previous code, have updated it now to the approach suggested, thanks!

sgiehl commented on plugins/UsersManager/Model.php:
```diff
@@ -342,7 +344,9 @@ public function addTokenAuth(
         $isSystemToken = false,
-        bool $secureOnly = false
+        bool $secureOnly = false,
+        bool $hasTokenExpiryDate = false,
+        string $tokenExpiryDate = null
```
What are those new parameters for? They seem to be unused.

michalkleiner replied:
I've added those when I started the work and then forgot to remove those as I found we already can handle it in the controller and we already had a column for the date. Nice catch, will remove.

sgiehl commented on plugins/UsersManager/vue/src/AddNewToken/AddNewToken.vue:
```diff
@@ -41,13 +47,51 @@
+          name="has_expiration"
+          :title="translate('UsersManager_TokenExpireDateCheckboxLabel')"
+          :required="false"
+          :inline-help="translate('UsersManager_TokenExpireDateCheckboxHelp')"
```
That one has a placehoder, which currently has no provided value.

michalkleiner replied:
good catch, added

sgiehl commented on plugins/UsersManager/vue/src/AddNewToken/AddNewToken.vue:
```diff
@@ -41,13 +47,51 @@
+            <input
+              type="text"
+              name="token_expire_date"
+              :value="tokenExpireDate"
```
wondering if it could make sense to have a required attribute here, that is set based on the value of `tokenHasExpiration`

michalkleiner replied:
good suggestion, added

sgiehl commented on plugins/UsersManager/Controller.php:
```diff
@@ -387,37 +387,67 @@ public function addNewToken()
 
+        $noDescription = null;
+        $description = '';
         if (!empty($_POST['description'])) {
```
As mentioned above: we should never directly access `$_POST`. Better to use `Request::fromPost`.

michalkleiner replied:
Updated to use Request::fromPost via a variable

sgiehl commented on plugins/UsersManager/Controller.php:
```diff
@@ -387,37 +387,67 @@ public function addNewToken()
-
             $description = \Piwik\Request::fromRequest()->getStringParameter('description', '');
+            $noDescription = false;
+        } elseif (isset($_POST['description'])) {
+            $noDescription = true;
+        }
```
Can't that be simplified in some way? If I see that correctly the variable is used to be passed through the template to vue.js. The vue component assumes the value to be passed as boolean, so passing the default `null` value, that is set, if the POST parameter is missing, might actually result in unexpected behavior.

michalkleiner replied:
fair point, updated and simplified

sgiehl commented on plugins/UsersManager/Controller.php:
```diff
@@ -387,37 +387,67 @@ public function addNewToken()
+            try {
+                $invalidExpireDate = true;
+                if (preg_match('/^(\d{4})-(\d{2})-(\d{2})$/', $tokenExpireDate)) {
+                    Date::factory($tokenExpireDate);
```
Maybe some check if the date is actually in the future might be helpful. Otherwise someone might by accident create a token that is already expired and would never show up in the list of tokens afterwards.

michalkleiner replied:
makes sense, added that check

sgiehl reviewed (changes requested):
Generally seems to work as expected. Left some comments on possible bugs and code improvements. Some additional tests might also be beneficial.

michalkleiner committed:
Apply initial review feedback

- require token expire date to be in future
- refactor and simplify controller logic
- remove unsude model method params
- always display expire column in the list of tokens
- pass default expiration to help text

michalkleiner committed:
Update config name to match emails-related work

michalkleiner committed:
Update UI test screenshot where expire date column is always visible

michalkleiner committed:
Update UI test screenshot with new global config options

michalkleiner committed:
Move initial expire date calculation to back-end

michalkleiner committed:
Use mock date for tests

michalkleiner committed:
Update UI tests

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Move logic to override Tests.now to the Date class

michalkleiner committed:
Merge branch '5.x-dev' into dev-19113-ui_api

michalkleiner committed:
Remove unused use statement

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Tweak Date logic around Tests.now

michalkleiner committed:
Fix datepicker visibility

caddoo commented on plugins/UsersManager/tests/Fixtures/ManyUsers.php:
```diff
@@ -162,4 +160,13 @@ protected function setUpUsers()
         $model->updateUserFields($this->pendingUser2['login'], ['invited_by' => 'login2']);
     }
+
+    public function provideContainerConfig()
+    {
+        return [
+            'Tests.now' => \Piwik\DI::decorate(function () {
+                return strtotime($this->dateTime); // not using Date::factory to prevent circular dependency
+            }),
+        ];
+    }
```
Just a comment, i know it's bad throughout our code that we don't inject a time instance, and rather just call time(), i think this direction is better overall, makes things a bit more testable, but not sure if it should be called Tests.now.

Just my thoughts, doesn't need changing now as it works. Just a naming thing.

michalkleiner committed:
Add UI test for token expiration date picker

caddoo commented on plugins/UsersManager/vue/src/AddNewToken/AddNewToken.vue:
```diff
@@ -41,13 +47,52 @@
+              for="token_expire_date"
+              class="active"
+            >{{ translate('UsersManager_TokenExpireDate') }}</label>
+            <input
```
As mention in call the input needs an ID for the label's for

michalkleiner replied:
ID added

michalkleiner committed:
Fix 'label for a missing id' accessibility issue

caddoo reviewed (commented):
Tested locally seems to work well.

Only small changes are the input ID so the label works.

But also a test to make sure the date picker appears would be appreciated.

@sgiehl if you are happy you can approve today. I am.

michalkleiner commented:
> But also a test to make sure the date picker appears would be appreciated.

UI test with the datepicker open added.

michalkleiner committed:
Use custom fixture not to impact other tests

michalkleiner committed:
Merge branch 'dev-19113' into dev-19113-ui_api

michalkleiner committed:
Fix CS

michalkleiner committed:
Fix fixture namespacing and use correct fixture in tests

michalkleiner committed:
Uncomment code commented out when debugging

michalkleiner committed:
Ensure initial date is set correctly for the date picker

michalkleiner committed:
Test config change to a different token expire interval

michalkleiner committed:
Update screenshot after config wording change

mneudert committed:
Remove unused function

mneudert committed:
Build vue components

mneudert committed:
Update expected screenshots

mneudert committed:
Remove obsolete comments

michalkleiner committed:
Allow numeric strings in Tests.now

michalkleiner committed:
Accept only values representing dates over 1990-01-01 in Tests.now

sgiehl committed:
set date correctly for tests
